rewriter/src/main/java/com/facebook/coresql/rewriter/CountDistinctRewriter.java

```diff
@@ -0,0 +1,122 @@
+
+    public CountDistinctRewriter(AstNode root)
+    {
+        this.root = requireNonNull(root, "AST passed to rewriter was null");
+        this.patternMatchedNodes = new CountDistinctPatternMatcher(root).matchPattern();
+    }
+
+    @Override
+    public Optional<RewriteResult> rewrite()
+    {
+        if (patternMatchedNodes.isEmpty()) {
+            return Optional.empty();
+        }
+        String rewrittenSql = unparse(root, this);
+        return Optional.of(new RewriteResult(REWRITE_NAME, rewrittenSql));
+    }
+
+    @Override
+    public void visit(Comparison node, Void data)
+    {
+        if (patternMatchedNodes.contains(node)) {
+            applyCountDistinctRewrite(node);
+        }
+        else {
+            defaultVisit(node, data);
+        }
+    }
+
+    private void applyCountDistinctRewrite(Comparison node)
+    {
+        // First, unparse up to the node. This ensures we don't miss any special tokens
+        unparseUpto(node);
+        // Then, add the rewritten version to the Unparser
+        String identifier = unparse(node.GetFirstChildOfKind(JJTAGGREGATIONFUNCTION).GetFirstChildOfKind(JJTIDENTIFIER)).trim();
+        printToken(format(REPLACEMENT, identifier, identifier));
+        // Move to end of this node -- we've already put in a rewritten version of it, so we don't need to unparse it
+        moveToEndOfNode(node);
+    }
+
+    private static class CountDistinctPatternMatcher
+            extends SqlParserDefaultVisitor
+    {
+        private final AstNode root;
+        private final ImmutableSet.Builder<AstNode> builder = ImmutableSet.builder();
+
+        public CountDistinctPatternMatcher(AstNode root)
+        {
+            this.root = requireNonNull(root, "AST passed to pattern matcher was null");
+        }
+
+        public Set<AstNode> matchPattern()
+        {
+            root.jjtAccept(this, null);
+            return builder.build();
+        }
+
+        private boolean secondArgIsLiteralOne(Comparison node)
+        {
+            Optional<AstNode> secondArg = Optional.ofNullable((AstNode) node.jjtGetChild(1));
+            return secondArg.isPresent() && unparse(secondArg.get()).trim().equals("1");
+        }
+
+        private boolean aggregationHasCountDistinct(AggregationFunction node)
+        {
+            Optional<AstNode> setQuantifier = Optional.ofNullable(node.GetFirstChildOfKind(JJTSETQUANTIFIER));
+            return setQuantifier.isPresent() && node.beginToken.image.equalsIgnoreCase("COUNT") && unparse(setQuantifier.get()).equalsIgnoreCase("DISTINCT");
+        }
+
+        private boolean isUsingCountDistinctComparisonToCheckUniqueness(Comparison node)
+        {
+            Optional<AstNode> aggregationFunction = Optional.ofNullable(node.GetFirstChildOfKind(JJTAGGREGATIONFUNCTION));
+            return aggregationFunction.isPresent() && aggregationHasCountDistinct((AggregationFunction) aggregationFunction.get()) && secondArgIsLiteralOne(node);
```

Comment:
We also need to check what the actual comparison is. If it's  = 1, then it can be rewritten to:
IF(MIN(x) IS NULL AND MAX(x) IS NULL, FALSE, MIN(x) = MAX(x))

if it's > 1,

IF(MIN(x) IS NULL AND MAX(x) IS NULL, FALSE, MIN(x) != MAX(x))

Reply:
We need to enhance the AST attributes to remember the operator. It's a TODO(viswanadha) :)

Reply:
Oops. I forgot I had the correct definition in the doc. So deleting these comments lol